### PR TITLE
Implement multipart partition object upload

### PR DESCRIPTION
**Pull request by gruuya**
Instead of loading up the entire file into memory and uploading it in one big chunk, perform a multipart upload of small buffered file chunks.

### CODE REVIEW

**Comment by gruuya** [comment]
As of the last implementation, the memory profile for the [dataset addition](https://www.splitgraph.com/docs/seafowl/getting-started/tutorial-fly-io/part-2-deploying-to-fly-io#add-the-dataset) from the docs (`13:08:21.452Z` is the start of the actual multipart partition files upload):
image

However, when running 
```sql
CREATE EXTERNAL TABLE area1 STORED AS PARQUET LOCATION '/home/ubuntu/area1.parquet';
CREATE TABLE area1 AS SELECT * FROM staging.area1;
```
via a psql terminal (using [area1.parquet](https://rdr.ucl.ac.uk/s/5cc49e8bcc7497581b30), with size about 2.45GB) I'm seeing:
image

File: src/context.rs [hunk 1]
```diff
@@ -304,6 +311,8 @@ pub async fn plan_to_object_store(
     }
     writer.close().map_err(DataFusionError::from).map(|_| ())?;
 
+    warn!("Starting upload of partition objects");
```
Comment (mildbyte) [inline src/context.rs:314]:
```suggestion
    info!("Starting upload of partition objects");
```

File: src/context.rs [hunk 2]
```diff
@@ -332,24 +341,57 @@ pub async fn plan_to_object_store(
...
+                                    break;
+                                } else if size != 0
+                                    && part_buffer.len() < PARTITION_FILE_MIN_PART_SIZE
+                                {
+                                    // Keep filling the chunk buffer until it surpasses the minimum required size
+                                    continue;
+                                }
+
+                                let part_size = part_buffer.len();
+                                warn!("Uploading part with {} bytes", part_size);
```
Comment (mildbyte) [inline src/context.rs:370]:
```suggestion
                                info!("Uploading part with {} bytes", part_size);
```

(or even debug?)

File: src/lib.rs [hunk 3]
```diff
@@ -1,3 +1,5 @@
+extern crate core;
+
```
Comment (mildbyte) [inline src/lib.rs:2]:
```suggestion
```

File: src/context.rs [hunk 4]
```diff
@@ -332,24 +341,57 @@ pub async fn plan_to_object_store(
...
+                    let (multipart_id, mut writer) =
+                        store.inner.put_multipart(&location).await?;
+
+                    let error: std::io::Error;
+                    loop {
+                        match reader.read_buf(&mut part_buffer).await {
+                            Ok(size) => {
+                                if size == 0 && part_buffer.is_empty() {
+                                    // We've reached EOF and there are no pending writes to flush
+                                    // TODO: as per the docs size = 0 doesn't actually guarantee that we've reached EOF
```
Comment (mildbyte) [inline src/context.rs:360]: Is this still true? Weird, since there's no other way to signal EOF (unless it returns a special Err type)?

**Comment by gruuya** [comment]
As a comparison, the current main show the following memory profiles for the [two examples above](https://github.com/splitgraph/seafowl/pull/99#issuecomment-1239393814):
**supply_chains**
image



**area1**
image

File: src/context.rs (hunk 4 above)
Comment (gruuya) [reply src/context.rs:360]: Yes, this is still true. I added a comment with a potential workaround using [stream_len](https://doc.rust-lang.org/nightly/std/io/trait.Seek.html#method.stream_len) (`nightly-only experimental API` atm).

I've also tried to get the true file size from metadata beforehand and keep track of the total read bytes, but they never matched somehow.

**Comment by gruuya** [comment]
Repeating profile once again (sanity check) for the current branch
**supply_chains**
image


**area1**
image

**Comment by gruuya** [comment]
Another memory profile, with upload task semaphore, and object store multipart race fix
**suply_chains**
image


**area1**
image

**Pull request merged**